check for conflicting revision and resolve with the most recent version

freefony opened the pull request:
closes #43

freefony committed:
check for conflicting revision and resolve with the most recent version

freefony committed:
feat: add conflict mangement to location module

tlvince commented on .idea/angular-nav-data.iml:
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Should you commit these IDE things?

tlvince commented on src/locations/locations.service.js:
```diff
@@ -1,4 +1,5 @@
 import { replication as replicationConfig } from '../config.json'
+import { default as utilService } from '../utils/utils.service'
```
As `checkAndResolveConflicts` is a pure function and doesn't rely on any class members in `utilService`, I'd move it out as a normal `export` or into a non-Angular wrapped file.

tlvince commented on src/locations/locations.service.js:
```diff
@@ -62,6 +63,8 @@ class LocationsService {
         .on('paused', onReplicationPaused)
     }
+
+    this.localDB.changes({conflicts: true, onChange: utilService.checkAndResolveConflicts.bind(null, this.localDB)})
```
Did you try without `bind`?

```diff
- this.localDB.changes({conflicts: true, onChange: utilService.checkAndResolveConflicts.bind(null, this.localDB)})
+ this.localDB.changes({conflicts: true, onChange: change => utilService.checkAndResolveConflicts(change, this.localDB)})
```

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,47 @@ class UtilsService {
+    if (!changedDoc.updatedAt) {
+      preferredRevision = changedDoc
+      changedDoc._conflicts.map((conflictingRev) => {
+        if (conflictingRev.updatedAt) {
```
Shouldn't `updateAt` be checked to see if it's newer than the currently chosen `preferredRevision`?

freefony replied:
the `if` checks that the `changedDoc` does not have `updatedAt` property so at this point `preferredRevision` does not have an `updatedAt` property so cannot be newer

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,47 @@ class UtilsService {
+    pouchdb.put(preferredRevision) // do we still need to PUT after plucking conflicts?
+  }
+
+  serialiseDocWithConflicts (Doc, prop) {
```
Why uppercase `Doc`? It's not a constructor.

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,47 @@ class UtilsService {
+  serialiseDocWithConflicts (Doc, prop) {
+    let mainDoc = {}
+    let serialisedDocs = []
+    Object.keys(Doc).map((prop) => {
```
Why iterate over all keys if you're only looking for `_conflicts`? Can't you do `if (Doc._conflicts) mainDoc._conflicts = Doc._conflicts`?

freefony replied:
Was trying to copy the doc without the `_conflicts` array property i fixed this in an update commit

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,47 @@ class UtilsService {
+        }
+      })
+    } else {
+      let serializedRevisions = this.serialiseDocWithConflictsByProp(changedDoc, 'updatedAt')
```
So this sorts conflicts by `updateAt`?

freefony replied:
Yes. also adds the main doc (couchdb's "winning" rev) so all are in an array and can be sorted

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,47 @@ class UtilsService {
+    serialisedDocs = [mainDoc]
+      .concat(Doc._conflicts)
+      .sort((a, b) => {
+        return a[prop] > b[prop]
```
This will always return `false` if `prop` (`updatedAt`) is `undefined`

tlvince reviewed (changes requested):
The logic seems straightforward enough, but I think we need some integration tests here to be confident it works as intended.

tlvince committed:
chore: add karma-env-preprocessor

tlvince committed:
test(locations): add integration test

tlvince committed:
fix(locations): use injected utils service

tlvince committed:
chore(travis): enable CouchDB for integration test

tlvince committed:
chore: rebundle

tlvince committed:
chore: rename to integration test

tlvince committed:
chore: delint

tlvince committed:
chore: check if DB exists

tlvince committed:
chore: enable CORS, randomise test DB

tlvince committed:
chore: clean up

tlvince committed:
chore: pin dev deps

tlvince committed:
chore(travis): test in node 6

tlvince committed:
chore: update babel and rollup

tlvince committed:
chore: remove unused uglify

tlvince committed:
chore: de-dupe helpers in bundle

tlvince committed:
chore: ignore yarn for now

tlvince committed:
chore: rebuild bundle

tlvince committed:
chore: remove focused spec now they're running

tlvince commented:
@freefony I took the liberty of writing a (failing) integration test. Please could you check the assertion and confirm whether my understand aligns with yours?

Also note I fixed a bug in the locations service whereby you need to use the injected version of the utils service (4886f2d). If you agree with that change, you'll also need to apply it to the other services.

(You can ignore the `chore` scopes above. They aren't directly related to this PR, but were necessary in order to run the integration test).

tlvince committed:
feat(locations): support restarting replication

tlvince committed:
test(locations): bump local rev

tlvince commented:
@freefony I've added restart replication support to locations service and made sure the local rev is at `2-`. However, so far it doesn't even look like the `onChange` handler is called.

tlvince committed:
fix(locations): correct change handler invocation

See:

* https://angular-pouchdb.github.io/angular-pouchdb/#/examples/changes
* https://pouchdb.com/api.html#changes

tlvince commented:
@freefony I fixed the change handler invocation in 3053a8f. Note angular-pouchdb's API is slightly different here.

freefony committed:
chore: fix slacks

tlvince commented:
The test still fails on `Error: undefined is not a constructor (evaluating 'this.serialiseDocWithConfl
ictsByProp(changedDoc, 'updatedAt')')`

tlvince commented:
That's fixed, but this is still not working as we'd expect as the `_conflicts` property is indeed a list of revs, not a list of docs. Note we're using PouchDB 5.4.x, not 6.x which the PouchDB docs are based on (might've been a source of confusion).

freefony commented:
Ok. i ll do `db.get` on each doc then

tlvince commented:
OK, NB, you don't need to fetch all rev docs by hand; you can use `open_revs: [<rev>] | 'all'` instead, e.g.:

```js
if (doc._conflicts.length) {
  db.get(doc._id, { open_revs: doc._conflicts })
  //=> [{}, {}, …]
}
```

Or indeed if there's only one conflict, use `rev: <rev>`.

tlvince commented on test/locations.service.integration.js:
```diff
@@ -17,7 +17,7 @@ function shouldNotBeCalled (rejection) {
 }
 
 describe('locationsService', function () {
-  var url = window.__env__.COUCHDB_URL || 'http://localhost:5984'
```
Was this causing any issues? I added it for ease of testing locally (can just run `npm t`)

tlvince reviewed (changes requested):
~~The test is failing for me with `Expected '2-' to be '3-'.`.~~

Actually, hold on, CouchDB wasn't started.

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,51 @@ class UtilsService {
+        let serializedRevisions = this.serialiseDocWithConflictsByProp(changedDoc, conflictingRevObjs, 'updatedAt')
+        preferredRevision = serializedRevisions.shift()
+        serializedRevisions.forEach(revision => {
+          pouchdb.remove(changedDoc._id, revision._rev)
```
Could you do this a `bulkDocs` perhaps?

tlvince commented on src/utils/utils.service.js:
```diff
@@ -59,6 +59,51 @@ class UtilsService {
+        })
+        preferredRevision._conflicts = []
+        pouchdb.put(preferredRevision)
+          .catch(err => {
```
Probably we should bubble these up to the caller so it can respond accordingly.

tlvince reviewed (changes requested):
The test is intermittently failing as `callOnReplicationComplete` occasionally resolves before `pouchdb.put(preferredRevision)`.

One way to solve this would be to add callback registration on the change handler similar to how's done for replication. That'd also be a good place to bubble any errors to.

tlvince commented:
@freefony I'm going to have look at this intermittent test.

tlvince committed:
test: restore default couch host

tlvince committed:
fix: add change callback, put winning rev

tlvince committed:
test: remove invalid assertion

This passed on PouchDB 6.1; suspect a PouchDB bug in pouchdb<6.1 (the version
pulled in by angular-pouchdb).

tlvince committed:
refactor: hide serialise from public API

tlvince committed:
chore: remove ignored IDE files

tlvince committed:
fix(products): dupe locations conflict handling

tlvince committed:
test: remove unneeded un registration

tlvince commented:
@freefony this lgtm now. Please have a run through a feel free to merge if it lgty 👍
